mappings/net/minecraft/block/Block.mapping

```diff
@@ -324,7 +324,7 @@ CLASS net/minecraft/class_2248 net/minecraft/block/Block
 		ARG 1 state
 		ARG 2 mirror
 	METHOD method_9570 hasBlockEntity ()Z
-	METHOD method_9571 (Lnet/minecraft/class_2680;Lnet/minecraft/class_1922;Lnet/minecraft/class_2338;)Lnet/minecraft/class_265;
+	METHOD method_9571 getShape (Lnet/minecraft/class_2680;Lnet/minecraft/class_1922;Lnet/minecraft/class_2338;)Lnet/minecraft/class_265;
```

Comment:
How does this differ from the other shape getters?

Reply:
From the commit message:
>  getShape. This delegates to the block's outline shape and is used to determine whether it is a full cube, and to get the culled shape.

It's a generic "getShape" method whereas the others are more specialised forms of the same method. i.e. `getOutlineShape()` and `getHitboxShape`. Presumably they did those so they can have blocks without the latter two but still have a shape to use where it's needed.

Reply:
This needs a better name. Look at the blocks that override it. Why does skull return empty and fence return `VoxelShapes.union(Block.createCuboidShape(7.0, 2.0, 0.0, 9.0, 13.0, 2.0), Block.createCuboidShape(7.0, 2.0, 14.0, 9.0, 13.0, 16.0));`? It's not just a "fallback shape".

Reply:
What would you call it instead then? I'll have to look more closely at the other methods to figure that out.

Reply:
The methods we have are:

```
    @Deprecated
    public VoxelShape getOutlineShape(BlockState state, BlockView view, BlockPos pos, EntityContext ePos) {
        return VoxelShapes.fullCube();
    }
    
    @Deprecated
    public VoxelShape getCollisionShape(BlockState state, BlockView view, BlockPos pos, EntityContext ePos) {
        return this.collidable ? state.getOutlineShape(view, pos) : VoxelShapes.empty();
    }
    
    @Deprecated
    public VoxelShape getShape(BlockState state, BlockView view, BlockPos pos) {
        return state.getOutlineShape(view, pos);
    }
    
    @Deprecated
    public VoxelShape getRayTraceShape(BlockState state, BlockView view, BlockPos pos) {
        return VoxelShapes.empty();
    }
```

The `getShape` method is overriden buy `FenceBlock`, `FenceGateBlock`, `LecternBlock`, and `SkullBlock`.

```
    // FenceBlock
    @Override
    public VoxelShape getShape(BlockState state, BlockView view, BlockPos pos) {
        return this.SHAPES[this.getShapeIndex(state)];
    }
    // FenceGateBlock
    @Override
    public VoxelShape getShape(BlockState state, BlockView view, BlockPos pos) {
        if (state.<Boolean>get((Property<Boolean>)FenceGateBlock.IN_WALL)) {
            return (state.<Direction>get((Property<Direction>)FenceGateBlock.FACING).getAxis() == Direction.Axis.X) ? FenceGateBlock.field_11027 : FenceGateBlock.field_11020;
        }
        return (state.<Direction>get((Property<Direction>)FenceGateBlock.FACING).getAxis() == Direction.Axis.X) ? FenceGateBlock.field_11023 : FenceGateBlock.field_11018;
    }
    // LecternBlock
    @Override
    public VoxelShape getShape(BlockState state, BlockView view, BlockPos pos) {
        return LecternBlock.BASE_SHAPE;
    }
    // SkullBlock
    @Override
    public VoxelShape getShape(BlockState state, BlockView view, BlockPos pos) {
        return VoxelShapes.empty();
    }
```

Reply:
@Runemoro @Juuxel After looking at how this method is being used, I decided to change the name to `getCullShape` (an overload of a similar method only present on `BlockState`)

This is based on what I found in `BlockState.ShapeCache` which is pre-computing the culled faces for each direction:
```
                this.culledFaces = new VoxelShape[BlockState.ShapeCache.DIRECTIONS.length];
                VoxelShape shape = block.getCullShape(state, EmptyBlockView.INSTANCE, BlockPos.ORIGIN);
                for (final Direction i : BlockState.ShapeCache.DIRECTIONS) {
                    this.culledFaces[i.ordinal()] = VoxelShapes.cull(shape, i);
                }
```

And also how these values, and the mapped method is being used here:

```
    public VoxelShape getCullShape(BlockView view, BlockPos pos, Direction facing) {
        if (this.shapeCache != null && this.shapeCache.culledFaces != null) {
            return this.shapeCache.culledFaces[facing.ordinal()];
        }
        return VoxelShapes.cull(this.getCullShape(view, pos), facing);
    }
```

Reply:
Interesting... Are you sure`VoxelShapes.cull` is named right though? To cull means to remove hidden faces, so if that method actually culls the shape, `BlockState.getCullShape` should actually be `getCulledShape`...

Reply:
> You're taking a cull shape, culling it, and obtaining another cull shape!?

Lol. You're taking a cull shape, culling it, and obtaining a _culled_ shaped.

Anyone is welcome to check my on that. I'm reasonably certain it has _something_ to do with culling. The method either returns `fullcube()`, `empty()`, or a `SlicedVoxelShape` cut along the direction's axis.

It seems like the slice is always going to have the face closest to the "observer" for lack of a better word.

```
public static VoxelShape cull(VoxelShape shape, Direction direction) {
		if (shape == fullCube()) {
			return fullCube();
		}
		Direction.Axis axis = direction.getAxis();
		boolean hasVolume = false;
		int sliceWidth = 0;
		
		if (direction.getDirection() == Direction.AxisDirection.POSITIVE) {
			hasVolume = DoubleMath.fuzzyEquals(shape.getMaximum(axis), 1, 1.0E-7);
			sliceWidth = shape.voxels.getSize(axis) - 1;
		} else {
			hasVolume = DoubleMath.fuzzyEquals(shape.getMinimum(axis), 0, 1.0E-7);
			sliceWidth = 0;
		}
		if (!hasVolume) {
			return empty();
		}
		return new SlicedVoxelShape(shape, axis, sliceWidth);
}
```

Reply:
> Lol. You're taking a cull shape, culling it, and obtaining a culled shaped.

Then why is the blockstate method *cull* and not *culled*?


But that doesn't matter anymore because `VoxelShapes.cull` is named wrong. That's not culling. The code you're showing takes a single face of the shape (its cross-section at the furthest point possible) and extrudes it to make it a prism width 1. The method should be named something like `extrudeFace` or `prismFromFace`. Its purpose is to convert the general cull shape into a side-specific cull shape.